packages/ui/aurora-grid/src/dnd/types.ts

```diff
@@ -0,0 +1,57 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { ForwardRefExoticComponent, HTMLAttributes, RefAttributes } from 'react';
+
+// Definitions:
+// - Mosaic: container that manages the layout of a set of tiles (e.g., Stack, Grid, Tree, Table).
+// - Tile: component that is rendered within the layout and is draggable.
+// - Item: datum represented by a Tile.
+
+export type MosaicDataItem = { id: string };
+
+export type MosaicDraggedItem = {
+  container: string;
+  item: MosaicDataItem;
+  position?: any; // Index or layout-specific positional information (stored separately from the item).
+};
+
+export type MosaicMoveEvent = {
+  active: MosaicDraggedItem;
+  over: MosaicDraggedItem;
+};
+
+/**
+ * props passed to mosaic tile.
+ */
+export type MosaicTileProps<TData extends MosaicDataItem> = Pick<HTMLAttributes<HTMLDivElement>, 'className'> & {
+  data: TData;
+
+  isActive?: boolean;
+  isDragging?: boolean;
+  draggableStyle?: any;
+  draggableProps?: any;
+  debug?: boolean;
+
+  // TODO(burdon): Generalize events (or use intents?)
+  onSelect?: () => void;
+};
+
+/**
+ * Tile component.
+ */
+export type MosaicTileComponent<TData extends MosaicDataItem> = ForwardRefExoticComponent<
+  RefAttributes<HTMLDivElement> & MosaicTileProps<TData>
+>;
+
+/**
+ * Tile container.
+ */
+export type MosaicContainerProps<TData extends MosaicDataItem> = {
+  id: string;
+  Component?: MosaicTileComponent<TData>;
+
+  // TODO(burdon): Rename drag events.
+  onMoveItem?: (event: MosaicMoveEvent) => void;
```

Comment:
you can try redefining the event type to have specific data and position types

example: https://github.com/dxos/dxos/blob/2f18e291944691b1af166a5af3e5712d417c5f4b/packages/core/halo/credentials/src/credentials/assertions.ts#L26

Reply:
let's have a conversation about generics. i think the general issue is:
the framework "routes" providers to consumers; both sides understand their own types -- but the framework has an `any` map. types are useful for casting, but the framework can't guarantee the type it's giving you. so isn't it just a cast?